File: services/moderation_service.py
import discord
from core.database import db
from typing import Optional, List, Dict
from datetime import datetime

class ModerationService:
    @staticmethod
    async def create_case(guild_id: int, target_id: int, moderator_id: int, action_type: str, reason: str, duration: Optional[str] = None) -> int:
        """Creates a new moderation case and returns the case number."""
        # Get next case number for this guild
        row = await db.fetch_one(
            "SELECT MAX(case_number) as last_case FROM moderation_cases WHERE guild_id = ?",
            guild_id
        )
        
        next_case = (row['last_case'] or 0) + 1
        
        await db.execute(
            """INSERT INTO moderation_cases 
               (guild_id, case_number, target_id, moderator_id, type, reason, duration) 
               VALUES (?, ?, ?, ?, ?, ?, ?)""",
            guild_id, next_case, target_id, moderator_id, action_type, reason, duration
        )
        
        return next_case

    @staticmethod
    async def get_user_cases(guild_id: int, target_id: int) -> List[Dict]:
        """Retrieves all cases for a specific user in a guild."""
        rows = await db.fetch_all(
            "SELECT * FROM moderation_cases WHERE guild_id = ? AND target_id = ? ORDER BY timestamp DESC",
            guild_id, target_id
        )
        return [dict(row) for row in rows]

    @staticmethod
    async def get_case(guild_id: int, case_number: int) -> Optional[Dict]:
        """Retrieves a specific moderation case."""
        row = await db.fetch_one(
            "SELECT * FROM moderation_cases WHERE guild_id = ? AND case_number = ?",
            guild_id, case_number
        )
        return dict(row) if row else None

    @staticmethod
    async def update_case(guild_id: int, case_number: int, **kwargs):
        """Updates a specific moderation case (e.g. for appeals)."""
        columns = ", ".join([f"{k} = ?" for k in kwargs.keys()])
        values = list(kwargs.values())
        
        query = f"UPDATE moderation_cases SET {columns} WHERE guild_id = ? AND case_number = ?"
        await db.execute(query, *values, guild_id, case_number)

    @staticmethod
    async def get_all_cases(guild_id: int, limit: int = 100) -> List[Dict]:
        """Retrieves a bulk list of cases for export."""
        rows = await db.fetch_all(
            "SELECT * FROM moderation_cases WHERE guild_id = ? ORDER BY case_number DESC LIMIT ?",
            guild_id, limit
        )
        return [dict(row) for row in rows]

    @staticmethod
    async def get_guild_config(guild_id: int) -> Optional[Dict]:
        """Retrieves configuration for a specific guild."""
        row = await db.fetch_one("SELECT * FROM guilds WHERE guild_id = ?", guild_id)
        return dict(row) if row else None

    @staticmethod
    async def update_guild_config(guild_id: int, **kwargs):
        """Updates or creates guild configuration."""
        columns = ", ".join([f"{k} = ?" for k in kwargs.keys()])
        values = list(kwargs.values())
        
        # Check if exists
        exists = await db.fetch_one("SELECT 1 FROM guilds WHERE guild_id = ?", guild_id)
        
        if exists:
            query = f"UPDATE guilds SET {columns} WHERE guild_id = ?"
            await db.execute(query, *values, guild_id)
        else:
            # Insert new
            cols = ["guild_id"] + list(kwargs.keys())
            placeholders = ", ".join(["?" for _ in cols])
            query = f"INSERT INTO guilds ({', '.join(cols)}) VALUES ({placeholders})"
            await db.execute(query, guild_id, *values)

File: src/services/moderationService.ts
import { Client, EmbedBuilder } from 'discord.js';
import { db } from '../core/database';
import logger from '../core/logger';

export class ModerationService {
    static async createCase(guildId: string, targetId: string, moderatorId: string, type: string, reason: string, duration?: string) {
        const row = await db.fetchOne(
            'SELECT MAX(case_number) as last_case FROM moderation_cases WHERE guild_id = ?',
            guildId
        );
        const nextCase = (Number(row?.last_case) || 0) + 1;

        await db.execute(
            'INSERT INTO moderation_cases (guild_id, case_number, target_id, moderator_id, type, reason, duration) VALUES (?, ?, ?, ?, ?, ?, ?)',
            guildId, nextCase, targetId, moderatorId, type, reason, duration ?? null
        );

        return nextCase;
    }

    static async getCase(guildId: string, caseNumber: number) {
        return db.fetchOne('SELECT * FROM moderation_cases WHERE guild_id = ? AND case_number = ?', guildId, caseNumber);
    }

    static async getUserHistory(guildId: string, userId: string) {
        return db.fetchAll('SELECT * FROM moderation_cases WHERE guild_id = ? AND target_id = ? ORDER BY case_number DESC', guildId, userId);
    }

    static async closeCase(guildId: string, caseNumber: number, note?: string) {
        await db.execute(
            'UPDATE moderation_cases SET case_status = ?, note = COALESCE(?, note) WHERE guild_id = ? AND case_number = ?',
            'closed', note ?? null, guildId, caseNumber
        );
    }

    static async addNote(guildId: string, caseNumber: number, note: string) {
        await db.execute(
            'UPDATE moderation_cases SET note = ? WHERE guild_id = ? AND case_number = ?',
            note, guildId, caseNumber
        );
    }

    static async getGuildConfig(guildId: string) {
        return db.fetchOne('SELECT * FROM guilds WHERE guild_id = ?', guildId);
    }

    static async setLogChannel(guildId: string, channelId: string) {
        const exists = await db.fetchOne('SELECT 1 FROM guilds WHERE guild_id = ?', guildId);
        if (exists) {
            await db.execute('UPDATE guilds SET log_channel_id = ? WHERE guild_id = ?', channelId, guildId);
        } else {
            await db.execute('INSERT INTO guilds (guild_id, log_channel_id) VALUES (?, ?)', guildId, channelId);
        }
    }

    // Post a formatted mod action to the guild's configured log channel
    static async sendLog(client: Client, guildId: string, embed: EmbedBuilder): Promise<void> {
        try {
            const config = await db.fetchOne('SELECT log_channel_id FROM guilds WHERE guild_id = ?', guildId);
            if (!config?.log_channel_id) return;

            const channel: any = await client.channels.fetch(config.log_channel_id).catch(() => null);
            if (!channel?.isTextBased()) return;

            await channel.send({ embeds: [embed] });
        } catch (err) {
            logger.warn(`Mod log send failed for guild ${guildId}: ${err}`);
        }
    }
}
